torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -321,13 +321,27 @@ IterDomain::IterDomain(
 
   TORCH_INTERNAL_ASSERT(
       _extent->isAnInt(),
-      "Cannot create an iter domain over an extent that is not an int but recieved ",
+      "Cannot create an iter domain over an extent that is not an int but received ",
       _extent,
       " .");
 
   TORCH_INTERNAL_ASSERT(
       _start->isAnInt(),
-      "Cannot create an iter domain with a start that is not an int but recieved ",
+      "Cannot create an iter domain with a start that is not an int but received ",
+      _extent,
+      " .");
+
+  // Check that all for-loops iterate from zero to some positive integer
+  // lower_insert_syncs uses this assumption for correctness.
+  TORCH_INTERNAL_ASSERT(
+      _start->isZeroInt(),
+      "Cannot create an iter domain with a start that is non-zero but received ",
+      _extent,
+      " .");
+
+  TORCH_INTERNAL_ASSERT(
+      !_extent->isZeroInt(),
```

Comment:
I don't think this always mean `_extent` is not zero. Does it only mean `_extent` is not known to be zero at the compile time.

Reply:
We could use the FusionExecutor to check the extent at runtime. However, it implies the user passed in zero for a symbolic value, which is a rare corner case.

Reply:
Yeah, that sounds too much, and I don't think we need to that. I was just wondering also because we may also insert predicates, so I wasn't sure whether we could say a loop body would be always executed at least once.

Reply:
Are there cases where we have an if-statement around a for-loop without an else-statement? Is this what you were referring to?

Reply:
I'm not sure. Another pattern may be a for loop whose body is predicated with a if statement. Again, I'm not sure we generate such code at this time.

I'm a little more concerned than usual since lack of proper synchronization may not be detected by just validating final results. That being said, I also feel I'm just too paranoid.

Reply:
We are appending the sync to the end of the for-loop and not the if-statement. The sync should run in this scenario.

As a general rule, we should run `cuda-memcheck --tool racecheck` on the shared memory unit tests and file race condition bugs if necessary.

Reply:
> We are appending the sync to the end of the for-loop and not the if-statement. The sync should run in this scenario.
> 

Right, but we also need to make sure that any further transformation after sync insertion must not do anything that could invalidate this assumption, but currently we have nothing to prevent that.